Fix wxLocale/wxUILocale issue #22281

utelle opened the pull request:
- Change wxLocale, so that always the default locale will be used (not the preferred UI language). This restores the behaviour of wxWidgets 3.1.4 and below.
- Add method GetSystemLocale to wxUILocale.
- Change the MSW implementation of wxUILocale, so that the default locale will be used instead of the preferred UI language for Windows versions below Windows 10.
- Change the Unix implementation of wxUILocale, so that the environment variable LANGUAGE is inspected for determining the preferred UI languages.
- Adjust wxTranslations to use wxUILocale to determine the preferred UI languages.
- Adjust documentation of wxUILocale.
- Adjust the internat sample to use wxUILocale during initialization.

utelle committed:
Fix issue 22281

- Change wxLocale, so that always the default locale will be used (not the preferred UI language). This restores the behaviour of wxWidgets 3.1.4 and below.
- Add method GetSystemLocale to wxUILocale.
- Change the MSW implementation of wxUILocale, so that the default locale will be used instead of the preferred UI language for Windows versions below Windows 10.
- Change the Unix implementation of wxUILocale, so that the environment variable LANGUAGE is inspected for determining the preferred UI languages.
- Adjust wxTranslations to use wxUILocale to determine the preferred UI languages.
- Adjust documentation of wxUILocale.
- Adjust the internat sample to use wxUILocale during initialization.

utelle committed:
Fix CI compile errors

utelle committed:
Add missing header file for logging

vadz commented on src/common/intl.cpp:
```diff
@@ -513,7 +513,7 @@ bool wxLocale::Init(int lang, int flags)
 int wxLocale::GetSystemLanguage()
 {
-    return wxUILocale::GetSystemLanguage();
+    return wxUILocale::GetSystemLocale();
```
I think this merits a comment saying that we intentionally use the "wrong" function here for compatibility, as otherwise the temptation to "fix" this by changing it to `GetSystemLanguage()` risks being too strong to resist.

utelle replied:
> I think this merits a comment saying that we intentionally use the "wrong" function here for compatibility, as otherwise the temptation to "fix" this by changing it to `GetSystemLanguage()` risks being too strong to resist.

Yes, you are right. I will add a comment.

vadz commented on src/common/intl.cpp:
```diff
@@ -679,6 +679,8 @@ void wxLocale::AddLanguage(const wxLanguageInfo& info)
 /* static */
 const wxLanguageInfo* wxLocale::GetLanguageInfo(int lang)
 {
+    if (lang == wxLANGUAGE_DEFAULT)
+        lang = GetSystemLanguage();
```
And the same thing here: we need to document that this is required, even if `wxUILocale` function already does the "same" thing internally because our `GetSystemLanguage()` is not the same.

In fact, maybe it would be better to use `wxUILocale::GetSystemLocale()` directly here, it could be more clear.

utelle replied:
> And the same thing here: we need to document that this is required, even if `wxUILocale` function already does the "same" thing internally because our `GetSystemLanguage()` is not the same.

Yes, for sure this should be documented.

> In fact, maybe it would be better to use `wxUILocale::GetSystemLocale()` directly here, it could be more clear.

I wanted to make clear that we use the "system language" as defined in the sense of `wxLocale`. If this _definition_ or its implementation would change, it could be easily forgotten to change both places - even if it is rather unlikely that the implementation of `wxLocale::GetSystemLanguage` changes once again.

vadz commented on src/common/translation.cpp:
```diff
@@ -96,11 +92,23 @@ wxStringToStringHashMap gs_msgIdCharset;
     wxLogTrace(TRACE_I18N, "%s: [%s]", prefix, wxJoin(arr, ','));
 }
 
+void LogTraceArray(const char *prefix, const wxVector<wxString>& arr)
+{
+    wxString s;
+    for (wxVector<wxString>::const_iterator j = arr.begin(); j != arr.end(); ++j)
+    {
+        if (j != arr.begin())
+            s += ",";
+        s += *j;
+    }
+    wxLogTrace(TRACE_I18N, "%s: [%s]", prefix, s);
+}
```
This is completely unrelated, but I wonder if we shouldn't add `wxJoin()` overload taking `wxVector<wxString>` instead, it looks like it could be useful.

And/or maybe change `wxJoin` to take `wxArrayStringsAdapter`?

utelle replied:
> This is completely unrelated, but I wonder if we shouldn't add `wxJoin()` overload taking `wxVector<wxString>` instead, it looks like it could be useful.

Implementing an additional `LogTraceArray` overload was the quickest way to get things compiled. Most likely it would be better to add a `wxJoin` overload.

> And/or maybe change `wxJoin` to take `wxArrayStringsAdapter`?

Sorry, never heard about nor used this class. It seems not to be documented.

vadz replied:
Sorry, this was mostly a note for myself, I'll check if I can do this later, this is not meant to prevent merging of this PR.

vadz commented on src/common/uilocale.cpp:
```diff
@@ -514,14 +514,14 @@ const wxUILocale& wxUILocale::GetCurrent()
+    }
+    else
+    {
+        m_impl = wxUILocaleImpl::CreateUserDefault();
```
I'm not sure if it's really a good idea to use the default locale for invalid `localeId`, why do we need to do this?

And if we do change it, the documentation for this ctor must be updated because it explicitly documents the current behaviour.

utelle replied:
> I'm not sure if it's really a good idea to use the default locale for invalid `localeId`, why do we need to do this?

We need it to restore the previous behaviour of `wxLocale` - selecting the system language based on the default locale (and not the preferred UI language).

> And if we do change it, the documentation for this ctor must be updated because it explicitly documents the current behaviour.

As said in my general comment, I simply forgot to update the documentation. However, we can revert the change and introduce a separate static method to instantiate a `wxUILocale` instance for the default locale.

vadz commented on src/common/uilocale.cpp:
```diff
@@ -664,8 +664,21 @@ int wxUILocale::GetSystemLanguage()
+/*static*/
+int wxUILocale::GetSystemLocale()
+{
+    // Retrieve default wxUILocale via empty wxLocaleIdent
+    wxLocaleIdent emptyLocId;
+    wxUILocale defaultLocale(emptyLocId);
+
```
If this is the only reason we need to change `wxUILocale` ctor above, I'd rather find a different solution, i.e. have a function returning default locale ID directly instead.

utelle replied:
> If this is the only reason we need to change `wxUILocale` ctor above, I'd rather find a different solution, i.e. have a function returning default locale ID directly instead.

Yes, that's the only reason. The default locale is needed here, but we have only method `wxUILocale::UseDefault` to create it - and that's not what we want to do here.

vadz commented on src/msw/uilocale.cpp:
```diff
@@ -32,6 +32,10 @@
 #endif
 
+#ifndef LOCALE_NAME_MAX_LENGTH
+#define LOCALE_NAME_MAX_LENGTH   85
```
Very minor, but for consistency we should do
```suggestion
    #define LOCALE_NAME_MAX_LENGTH   85
```

vadz commented on src/msw/uilocale.cpp:
```diff
@@ -415,19 +422,28 @@ class wxUILocaleImplName : public wxUILocaleImpl
-            ULONG numberOfLanguages = 0;
-            ULONG bufferSize = 0;
-            if (ms_GetUserPreferredUILanguages(MUI_LANGUAGE_NAME, &numberOfLanguages, NULL, &bufferSize))
+            // Check for Windows version >= 10
+            // Note: Windows 8.x might be supported here as well, but Windows 7 and below definitely not
```
```suggestion
            // Check if Windows supports preferred UI languages.
            // Note: Windows 8.x might support them as well, but Windows 7
            // and below definitely do not.
```

vadz commented on src/msw/uilocale.cpp:
```diff
@@ -437,7 +453,18 @@ class wxUILocaleImplName : public wxUILocaleImpl
+                {
+                    wxLogLastError(wxT("GetUserDefaultLocaleName"));
+                }
+
```
```suggestion
```

vadz commented on src/msw/uilocale.cpp:
```diff
@@ -437,7 +453,18 @@ class wxUILocaleImplName : public wxUILocaleImpl
+                wchar_t buf[LOCALE_NAME_MAX_LENGTH];
+                if (!ms_GetUserDefaultLocaleName(buf, LOCALE_NAME_MAX_LENGTH))
+                {
+                    const wxString language(buf);
+                    preferred.push_back(language);
```
Do we really need a temporary variable? I.e. why not just
```suggestion
                    preferred.push_back(buf);
```

vadz reviewed (commented):
Thanks a lot!

Globally this looks good to me, but I'd like to avoid changing `wxUILocale` ctor if possible -- or understand why do we need to change it.

And I think a few comments should be added for future-proofing.

utelle commented:
> Globally this looks good to me, but I'd like to avoid changing `wxUILocale` ctor if possible -- or understand why do we need to change it.

The only change I made to the `wxUILocale` ctor is that passing an empty `wxLocaleIdent` (for the default locale) is now allowed and results in creating an instance of the default locale . This is a simple way to retrieve the default locale. However, if you prefer, we can introduce an additional static convenience method to explicitly return an instance of the default locale.

> And I think a few comments should be added for future-proofing.

Sorry, I forgot to update the documentation of the `wxUILocale` ctor. I'll catch up on that, unless we decide to revert to the old behaviour and to introduce a separate method. Where else are comments missing?

utelle committed:
Update src/msw/uilocale.cpp

Remove temporary variable

Co-authored-by: VZ <vz-github@zeitlins.org>

utelle committed:
Update src/msw/uilocale.cpp

Co-authored-by: VZ <vz-github@zeitlins.org>

utelle committed:
Update src/msw/uilocale.cpp

Update comment for check of Windows version

Co-authored-by: VZ <vz-github@zeitlins.org>

utelle committed:
Update src/msw/uilocale.cpp

Co-authored-by: VZ <vz-github@zeitlins.org>

vadz commented:
I'd indeed prefer to have a separate method (possibly private because I don't see why would people need to construct the default locale without using it) for this. The rationale is that we want to allow using `wxUILocale(locId).IsSupported()` to work for checking if the locale is supported, but now this would always return true. In fact, AFAICS, there would be no way to create an invalid object of this class at all.

utelle commented:
> I'd indeed prefer to have a separate method (possibly private because I don't see why would people need to construct the default locale without using it) for this.

I can't see how we could make this method private, because we need to call it from a `wxLocale` method.

> The rationale is that we want to allow using `wxUILocale(locId).IsSupported()` to work for checking if the locale is supported, but now this would always return true.

Why do you believe that?

Under Windows asking for a locale for "de-US" will fail since it is constructed (and our code rejects it); under macOS we check whether the requested locale is in the list of available locales (and reject it, if not); under Unix the function `newlocale` will fail, if a non-existing locale id is used.

That is, for all platforms invalid locale ids will be rejected, and `wxUILocale::IsSupported` will then return `false`.

I only introduced a "special" id (namely an empty instance of `wxLocaleIdent`) for the default locale. Just like having a language id `wxLANGUAGE_DEFAULT`. Usually user code will use `wxLocaleIdent::Language()` to set at least the language code, and can check afterwards with method `wxLocaleIdent::IsEmpty()` whether the language was actually set.

> In fact, AFAICS, there would be no way to create an invalid object of this class at all.

As explained above this is simply not true.

IMHO using an empty `wxLocaleIdent` for the default locale attracts less attention than introducing a separate method (which can't be made private, unless we make `wxLocale` a friend of `wxUILocale`). Or do you mean with "private" that we implement a separate _function_ that is only defined in the header `private/uilocale.h` and not documented?

vadz commented:
> > I'd indeed prefer to have a separate method (possibly private because I don't see why would people need to construct the default locale without using it) for this.
> 
> I can't see how we could make this method private, because we need to call it from a `wxLocale` method.

We should be able to set things up so that it could be done in a private function declared in `wx/private/uilocale.h` and used inside `wxUILocale` as well, I think.

> > The rationale is that we want to allow using `wxUILocale(locId).IsSupported()` to work for checking if the locale is supported, but now this would always return true.
> 
> Why do you believe that?

Oh, sorry, it won't always return true, of course, you're right and I was completely wrong to write that. What I meant (but definitely not what I wrote, sorry again) is that we won't be able to use this construct any more to verify that `locId` is valid -- now it would mean that it's valid or empty (which, of course, means invalid at `wxLocaleIdent` level), which doesn't seem right.
 
> IMHO using an empty `wxLocaleIdent` for the default locale attracts less attention than introducing a separate method (which can't be made private, unless we make `wxLocale` a friend of `wxUILocale`). Or do you mean with "private" that we implement a separate _function_ that is only defined in the header `private/uilocale.h` and not documented?

Yes, this is exactly what I meant. "Private" for me means "not part of public API", not necessarily "private C++ function".

utelle committed:
Restore wxUILocale ctor

- Implement a simpler method to create an instance of the default localeconv
- Add some clarifying comments in wxLocale implementation

utelle commented:
> > I can't see how we could make this method private, because we need to call it from a `wxLocale` method.
> 
> We should be able to set things up so that it could be done in a private function declared in `wx/private/uilocale.h` and used inside `wxUILocale` as well, I think.

Actually, we don't need any additional helper function/method. I simply missed that a static public method can call internally private methods.

> Yes, this is exactly what I meant. "Private" for me means "not part of public API", not necessarily "private C++ function".

Sorry for the noise. I just committed a variant that gets away without a private helper function.

I also added comments in the wxLocale methods, and restored the previous `wxUILocale` ctor. That is, this PR should hopefully now be ready for merging.

vadz commented:
Thanks! Looks good to me and I'll (squash) merge this soon but will try to fix the Mac CI builds first to avoid more failures in them.

There also seems to be an accidental change to the libtiff submodule here, I won't apply it (nor some whitespace-only changes w.r.t. master).

Thanks again!

utelle commented:
> There also seems to be an accidental change to the libtiff submodule here, I won't apply it (nor some whitespace-only changes w.r.t. master).

No idea how this libtiff change slipped in, I didn't change anything consciously.

Thanks for applying the PR.

vadz commented:
You probably switched between your branch and master and didn't update the submodule (this isn't done automatically when switching branches) and finally committed everything (i.e. `git commit -a`) including the locally changed (because not yet updated) submodule.

FWIW I have the following in my `.git/hooks/pre-commit`:
```sh
if [ "$SUBMODULES_OK" != 1 ]; then
    if git diff-index --cached HEAD | grep --quiet '^:160000'; then
        echo
        echo "Error: Not committing submodules, use SUBMODULES_OK=1 to allow"
        echo
        exit 1
    fi
fi
```
to prevent myself from committing submodules accidentally.

utelle commented:
> You probably switched between your branch and master and didn't update the submodule

That's it! Thanks for the hint.

> FWIW I have the following in my `.git/hooks/pre-commit`: [...]
> to prevent myself from committing submodules accidentally.

I will add this pre-commit script to my local git repository. Thanks a lot!
